lib/ansible/plugins/connection/ssh.py

```diff
@@ -309,8 +309,49 @@ def _examine_output(self, source, state, b_chunk, sudoable):
 
         return b''.join(output), remainder
 
+
     def _run(self, cmd, in_data, sudoable=True):
         '''
+        Will retry execution if ssh were unable to connect or timeout
+        '''
+        remaining_tries = int(C.ANSIBLE_SSH_RETRIES) + 1
+        cmd_summary = ' '.join(cmd)
+        for attempt in range(remaining_tries):
+            try:
+                return_tuple = self._run_once(cmd=cmd, in_data=in_data, sudoable=sudoable)
+                # 0 = success
+                # 1-254 = remote command return code
+                # 255 = failure from the ssh command itself
+                if return_tuple[0] != 255:
+                    break
+                else:
```

Comment:
I have a branch that moves the return code check into _run_once, and removes some of the nesting mentioned above. https://github.com/ansible/ansible/compare/devel...alikins:ssh_error_raise_in_run_once?expand=1